From 08e5128ffd63acd14249db3930f05753abb8f848 Mon Sep 17 00:00:00 2001
From: Adam Gandelman <adamg@canonical.com>
Date: Mon, 23 Jul 2012 13:16:46 -0700
Subject: [PATCH] Backport fix for API listing of os-hosts.

Backports fix for bug 1014925 to stable/essex, which resolves issue
where querying /v1.1/$tenant/os-hosts returns an empty list.

Original fix by Joe Gordon reviewed into Folsom at:

    https://review.openstack.org/#/c/8682/2

Change-Id: I44ac2e519b7af9b8f3b37a42280ac6fe71c31a1c
---
 nova/api/openstack/compute/contrib/hosts.py        |    6 ++++--
 .../api/openstack/compute/contrib/test_hosts.py    |   11 ++++++++---
 2 files changed, 12 insertions(+), 5 deletions(-)

diff --git a/nova/api/openstack/compute/contrib/hosts.py b/nova/api/openstack/compute/contrib/hosts.py
index a93da9e..202c8ca 100644
--- a/nova/api/openstack/compute/contrib/hosts.py
+++ b/nova/api/openstack/compute/contrib/hosts.py
@@ -27,7 +27,6 @@ from nova import db
 from nova import exception
 from nova import flags
 from nova import log as logging
-from nova.scheduler import api as scheduler_api
 
 
 LOG = logging.getLogger(__name__)
@@ -98,7 +97,10 @@ def _list_hosts(req, service=None):
     by service type.
     """
     context = req.environ['nova.context']
-    hosts = scheduler_api.get_host_list(context)
+    services = db.service_get_all(context, False)
+    hosts = []
+    for host in services:
+        hosts.append({"host_name": host['host'], 'service': host['topic']})
     if service:
         hosts = [host for host in hosts
                 if host["service"] == service]
diff --git a/nova/tests/api/openstack/compute/contrib/test_hosts.py b/nova/tests/api/openstack/compute/contrib/test_hosts.py
index 77beeae..0482eb5 100644
--- a/nova/tests/api/openstack/compute/contrib/test_hosts.py
+++ b/nova/tests/api/openstack/compute/contrib/test_hosts.py
@@ -36,10 +36,15 @@ HOST_LIST = [
         {"host_name": "host_c2", "service": "compute"},
         {"host_name": "host_v1", "service": "volume"},
         {"host_name": "host_v2", "service": "volume"}]
+SERVICES_LIST = [
+        {"host": "host_c1", "topic": "compute"},
+        {"host": "host_c2", "topic": "compute"},
+        {"host": "host_v1", "topic": "volume"},
+        {"host": "host_v2", "topic": "volume"}]
 
 
-def stub_get_host_list(req):
-    return HOST_LIST
+def stub_service_get_all(self, req):
+    return SERVICES_LIST
 
 
 def stub_set_host_enabled(context, host, enabled):
@@ -104,7 +109,7 @@ class HostTestCase(test.TestCase):
         super(HostTestCase, self).setUp()
         self.controller = os_hosts.HostController()
         self.req = FakeRequest()
-        self.stubs.Set(scheduler_api, 'get_host_list', stub_get_host_list)
+        self.stubs.Set(db, 'service_get_all', stub_service_get_all)
         self.stubs.Set(self.controller.api, 'set_host_enabled',
                        stub_set_host_enabled)
         self.stubs.Set(self.controller.api, 'set_host_maintenance',
